docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,775 @@
+    might require configuration, which is why there is a `config_data`
+    parameter. The configuration data can be plugin specific, so no format is
+    specified in this proposal.
+- `get_evidence`, `free_evidence`, and `free_endorsements`
+  - Producing evidence and endorsements is necessary for attestation.
+  - `flags` field to determine local vs. remote attestation.
+  - There is a `custom_claims` parameter because many attestation protocols
+    require the enclave to sign some data from a relying party. For example,
+    many protocols follow the "challenge response" architecture, which requires
+    the enclave to sign a nonce from the relying party.
+  - There is an `opt_params` field because some plugins might require plugin
+    specific input. For example, the SGX local attestation needs the
+    other enclave's target info struct.
+  - There is an `endorsements` parameter to return the endorsements that are
+    coupled with the evidence to ensure that the evidence and endorsements are
+    in sync.
+- `verify_evidence`
+  - Verifying evidence and endorsements is essential for attestation.
+  - The `claims` field contains key-value pairs that can be verified by the
+    caller. This will have the similar contents as the `oe_identity_t` field
+    in the `oe_report_t` struct returned by `oe_verify_report` and any custom
+    claims that were passed to the `get_evidence` function.
+  - The `input_validation_time` is needed to validate the evidence against a
+    specific time.
+
+###  Known Open Enclave Claims
+
+- Each plugin's `verify_evidence` function must, at minimum, return the
+  following claims (mapped from the `oe_identity_t`):
+  
+| Claim Name       | Claim Value Type   | Description                                                          |
+|:-----------------|:-------------------|:---------------------------------------------------------------------|
+| id_version       | uint32_t           | Claims version. Must be 0                                            |
+| security_version | uint32_t           | Security version of the enclave. (ISVN for SGX).                     |
+| attributes       | uint64_t           | Attributes flags for the evidence: <br/> `OE_REPORT_ATTRIBUTES_DEBUG`: The evidence is for a debug enclave.<br/> `OE_REPORT_ATTRIBUTES_REMOTE`: The evidence can be used for remote attestation.   |
+| unique_id        | uint8_t[32]        | The unique ID for the enclave (MRENCLAVE for SGX).                   |
+| signer_id        | uint8_t[32]        | The signer ID for the enclave (MRSIGNER for SGX).                    |
+| product_id       | uint8_t[32]        | The product ID for the enclave (ISVPRODID for SGX).                  |
+| validity_from    | oe_datetime_t      | Overall datetime from which the evidence and endorsements are valid. |
+| validity_until   | oe_datetime_t      | Overall datetime at which the evidence and endorsements expire.      |
```

Comment:
Another thing I would like to ask about. With SGX, there is only one TA to provide evidence for. With OPTEE, there are multiple levels of firmware, each with its own set of claims. We are currently expressing these as a [caller in oe_identity_t](https://github.com/openenclave/openenclave/blob/feature.new_platforms/include/openenclave/bits/report.h). Wondering if for example capturing claims about Quoting Enclave for SGX would be interesting.

Reply:
I could see how they align. The qeidentity field in the quote essentially all the `oe_identity_t` information of the quoting enclave. In both SGX and CyRes, it seems that the caller field can be the parent component of the TA in the trust chain.

Reply:
It would be awesome if we could align that. If you look at the `feature.new_platforms` branch, I have made some changes for OPTEE wrt the APIs as well (you need to free the stuff now). Please have a look if that is something that could be adopted.